Adding initial draft of `ww-proseq` WILDS WDL pipeline (#337)
getwilds/wilds-wdl-library: pipelines/ww-proseq/README.md

   - `spikein_chrom_prefix` — the prefix used on spike-in contig names in the merged FASTA
3. **Pipeline parameters** — UMI config, MAPQ threshold, resource allocation

See [`inputs.json`](inputs.json) for a template.

### Building the merged spike-in reference

The `ww-testdata.merge_fastas_with_prefix` task can build the merged FASTA programmatically:

```wdl
call ww_testdata.merge_fastas_with_prefix as build_merged { input:
  first_fasta = experimental_fasta,
  second_fasta = spikein_fasta,
  second_prefix = "hg38",
  output_name = "exp_spikein_merged"
}
```

Or build it offline once with:
```bash
cat experimental.fa <(sed 's/^>/>hg38/' spikein.fa) > merged.fa
samtools faidx merged.fa
```

### Running the Pipeline

```bash
# Using miniWDL
miniwdl run ww-proseq.wdl --input inputs.json

# Using Sprocket
sprocket run ww-proseq.wdl --inputs inputs.json --entrypoint proseq

# Using Cromwell
java -jar cromwell.jar run ww-proseq.wdl --inputs inputs.json
```

## Input Parameters

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `samples` | `Array[ProseqSample]` | required | Per-sample name + R1 + R2 |

> | `references` | `ProseqReferences` | required | Three FASTAs + spike-in chromosome prefix |

Reviewer: ```suggestion
| `references` | `ProseqReferences` Struct | required | Three FASTAs + spike-in chromosome prefix |
```

> | `references` | `ProseqReferences` | required | Three FASTAs + spike-in chromosome prefix |

Reviewer: ```suggestion
| `samples` | `Array[ProseqSample Struct]` | required | Per-sample name + R1 + R2 |
| `references` | `ProseqReferences Struct` | required | Three FASTAs + spike-in chromosome prefix |
```
I would clarify what these are

Author: You bring up a good point that we need to clarify what these structs look like. Added a link in the "description" column and added more detail to those sections in https://github.com/getwilds/wilds-wdl-library/pull/337/commits/ba9280fc330ed32ff5742c14c2d3215266a15a7e

| `align_memory_gb` | `Int` | `8` | Memory per bowtie2 alignment task |

### `ProseqSample` Structure

```wdl
struct ProseqSample {
    String name
    File r1
    File r2
}
```

### `ProseqReferences` Structure

```wdl
struct ProseqReferences {
    File experimental_fasta
    File spikein_merged_fasta
    File rdna_fasta
    String spikein_chrom_prefix
}
```

## Output Files

Per sample (arrays scattered over `samples`):
- `fastp_r1_trimmed`, `fastp_r2_trimmed` — UMI-extracted, adapter-trimmed FASTQs
- `fastp_html`, `fastp_json` — fastp reports
- `experimental_bam_dedup`, `experimental_bai_dedup` — deduplicated experimental BAM
- `experimental_dedup_log` — `umi_tools dedup` log (input/output read counts, PCR-dup stats)
- `spikein_bam_dedup`, `spikein_bai_dedup` — deduplicated spike-in BAM (for cross-sample normalization)
- `spikein_dedup_log` — `umi_tools dedup` log for the spike-in BAM
- `bigwig_forward`, `bigwig_reverse` — strand-specific 3' single-base bigWigs

Aggregated:
- `multiqc_report` — MultiQC HTML
- `multiqc_data` — MultiQC parsed-metrics directory

## Resource Considerations

### Compute Requirements

| Step | CPU | Memory | Notes |
|------|-----|--------|-------|

> | `bowtie2_build` × 3 | 4 cores | 16 GB | Once per pipeline run |

Reviewer: ```suggestion
| `bowtie2_build` × 3 | 4 cores | 16 GB | Runs once per pipeline run for each reference FASTA |
```
If this is correct I would include to clarify why the x3

| `bowtie2_align` × 3 per sample | 4-8 cores | 8-16 GB | Memory scales with reference size, especially the spike-in-merged reference |
| `umi_tools dedup` × 2 per sample | 2 cores | 8 GB | Single-threaded; memory scales with read density |

> | `bam_coverage` × 2 per sample | 4 cores | 8 GB | Two strand-specific bigWigs per sample |

Reviewer: ```suggestion
| `bam_coverage` × 2 per sample | 4 cores | 8 GB | Runs on two strand-specific bigWigs per sample |
```
If this is correct I would include to clarify what is meant by this sentence

### Optimization Tips

- The spike-in alignment is run against the **merged** reference (experimental + spike-in concatenated), then filtered. This matches the original PROseq_alignment.sh script and ensures spike-in normalization factors are computed on reads that are unambiguously spike-in.
- For very deep PRO-seq runs (>100M reads/sample), bump `align_memory_gb` to 32 GB.

> - The bigWig step is the only place that does any biological math — `--Offset 1` is what makes this PRO-seq specific (3' read end = active site of Pol II).

Reviewer: ```suggestion
- The `bam_coverage` bigWig step is the only place that does any biological math — `--Offset 1` is what makes this PRO-seq specific (3' read end = active site of RNA Polymerase II).
```
If correct I would include these clarifying details

## Testing the Pipeline

The `testrun.wdl` is zero-config:

```bash
# Using miniWDL
miniwdl run testrun.wdl

# Using Sprocket
sprocket run testrun.wdl --entrypoint proseq_example
```

It:
1. Pulls 50K read pairs from SRA accession **SRR11607571** (Judd et al. 2020 qPRO-seq, Drosophila S2 cells — same lab/protocol as the source script)
2. Downloads small dm6 (chr2L:1-1Mb) and hg38 (chr1:1-1Mb) reference fragments via `ww-testdata.download_ref_data`
3. Builds the merged spike-in reference via `ww-testdata.merge_fastas_with_prefix`
4. Pulls the human 45S rRNA precursor via `ww-testdata.download_rrna_reference`
5. Runs the full pipeline end-to-end on the synthetic-but-realistic dataset

> Total testrun duration: ~10-15 minutes (dominated by the SRA download).

Reviewer: 10-15 min on what system?

Author: Now that I look at it, it's a little less than 5 on the GitHub runners, updated to reflect that in https://github.com/getwilds/wilds-wdl-library/pull/337/commits/248b30533f3b22f2dc819df588056f9a45381079

## Citation

> If you use this pipeline, please cite the qPRO-seq protocol it implements:

Reviewer: I think we need to pick one citation for people to use, ideally the most 'original' one. Otherwise include a link to the tool/pipeline homepage and let the user figure out the proper citation themselves. This is too many things considering we aren't the authors

Author: Totally fair, I kinda overloaded that section, figured it was safer to add more than less, but you're right, it's a bit too crowded. Reduced to the original paper in https://github.com/getwilds/wilds-wdl-library/pull/337/commits/248b30533f3b22f2dc819df588056f9a45381079
